Fails on UNICODE builds on Windows

Compiling with the `UNICODE` macro (which is required by most other well-behaving Windows libraries) currently fails.

If you want to be ASCII-only then you must explicitly call the `*A` versions of functions, e.g. `FindFirstFileA` instead of `FindFirstFile`. The unadorned versions are macros that expand to either `FindFirstFileW` or `FindFirstFileA` depending on whether `_UNICODE` is in use or not.

A better bet would be to just be compatible with `_UNICODE` as there are some places where you need Windows' native UTF16 paths and can't use the ASCII variants. To do this you need to use `TCHAR` instead of `char` for your strings and buffers and use the the `TEXT` macro to wrap all your literals.

Yeah, Windows is lame for not just being UTF8 everywhere, but it is what it is. Recommended practice for Win32 developers is to use UTF8 internally for all strings, enable `_UNICODE` to catch mistakes like this one with Win32 calls, to always explicitly call the *W variants of functions (in case a user turns off `_UNICODE`), and then do any translation to and from UTF16/UTF8 at the boundaries where Win32 OS calls are made.

For a library like this one, I think the best option would be to just support both modes using the macros described above.

Errors received with latest tinydir when using `_UNICODE`:

    1>C:\Users\Sean\Documents\Projects\tinydir/tinydir.h(142): error C2664: 'HANDLE FindFirstFileW(LPCWSTR,LPWIN32_FIND_DATAW)' : cannot convert argument 1 from 'char [4096]' to 'LPCWSTR'
    1>          Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
    1>C:\Users\Sean\Documents\Projects\tinydir/tinydir.h(327): error C2664: 'size_t strlen(const char *)' : cannot convert argument 1 from 'const WCHAR [260]' to 'const char *'
    1>          Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
    1>C:\Users\Sean\Documents\Projects\tinydir/tinydir.h(340): error C2664: 'size_t strlen(const char *)' : cannot convert argument 1 from 'const WCHAR [260]' to 'const char *'
    1>          Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
    1>C:\Users\Sean\Documents\Projects\tinydir/tinydir.h(354): error C2664: 'char *strcpy(char *,const char *)' : cannot convert argument 2 from 'const WCHAR [260]' to 'const char *'
    1>          Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast

--- a/samples/list_to_file.c
+++ b/samples/list_to_file.c
@@ -2,6 +2,7 @@
 
 #ifndef TINYDIR_H
 #define _UNICODE
+#define UNICODE
 #include "tinydir.h"
 #endif
 

--- a/samples/list_to_file.c
+++ b/samples/list_to_file.c
@@ -5,6 +5,8 @@
 #include "tinydir.h"
 #endif
 
+char bom[] = "\xFF\xFE";
+
 int main(void) {
 
   FILE *fp;
@@ -14,7 +16,6 @@ int main(void) {
   fp = fopen("/file/to/output","wb");
 
 #ifdef _WIN32
-  char bom[] = "\xFF\xFE";
   fwrite(bom, 1, 2, fp);
 #endif
 

--- a/samples/list_to_file.c
+++ b/samples/list_to_file.c
@@ -13,7 +13,13 @@ int main(void) {
   tinydir_dir dir;
   tinydir_open(&dir, TINYDIR_STRING("/path/to/dir"));
 
-  fp = fopen("/file/to/output","wb");
+  fp = 
+#ifdef _WIN32
+	_wfopen(
+#else
+	fopen(
+#endif
+		TINYDIR_STRING("/file/to/output"), TINYDIR_STRING("wb"));
 
 #ifdef _WIN32
   fwrite(bom, 1, 2, fp);

--- a/samples/list_to_file.c
+++ b/samples/list_to_file.c
@@ -1,0 +1,41 @@
+﻿#include <stdio.h>
+
+#ifndef TINYDIR_H
+#define _UNICODE
+#include "tinydir.h"
+#endif
+
+int main(void) {
+
+  FILE *fp;
+  tinydir_dir dir;
+  tinydir_open(&dir, TINYDIR_STRING("/path/to/dir"));
+
+  fp = fopen("/file/to/output","wb");
+
+#ifdef _WIN32
+  char bom[] = "\xFF\xFE";
+  fwrite(bom, 1, 2, fp);
+#endif
+
+  while (dir.has_next)
+  {
+    tinydir_file file;
+    tinydir_readfile(&dir, &file);
+
+    fwrite(file.name, sizeof(_tinydir_char_t), _tinydir_strlen(file.name), fp);
+    if (file.is_dir)
+    {
+        fwrite(TINYDIR_STRING("/"), sizeof(_tinydir_char_t), _tinydir_strlen(TINYDIR_STRING("/")), fp);
+    }
+    fwrite(TINYDIR_STRING("\n"), sizeof(_tinydir_char_t), _tinydir_strlen(TINYDIR_STRING("/")), fp);
+
+    tinydir_next(&dir);
+  }
+
+  tinydir_close(&dir);
+
+  fclose(fp);
+
+  return 0;
+}

--- a/samples/list_to_file.c
+++ b/samples/list_to_file.c
@@ -14,14 +14,14 @@ int main(void) {
   tinydir_open(&dir, TINYDIR_STRING("/path/to/dir"));
 
   fp = 
-#ifdef _WIN32
+#if ((defined _WIN32) && (defined _UNICODE))
 	_wfopen(
 #else
 	fopen(
 #endif
 		TINYDIR_STRING("/file/to/output"), TINYDIR_STRING("wb"));
 
-#ifdef _WIN32
+#if ((defined _WIN32) && (defined _UNICODE))
   fwrite(bom, 1, 2, fp);
 #endif
 

--- a/tinydir.h
+++ b/tinydir.h
@@ -32,6 +32,14 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 extern "C" {
 #endif
 
+#if ((defined _UNICODE) & !(defined UNICODE))
+#define UNICODE
+#endif
+
+#if ((defined UNICODE) & !(defined _UNICODE))
+#define _UNICODE
+#endif
+
 #include <errno.h>
 #include <stdlib.h>
 #include <string.h>

--- a/tinydir.h
+++ b/tinydir.h
@@ -52,9 +52,9 @@ extern "C" {
 /* types */
 
 /* Windows UNICODE wide character support */
-#if ((defined _MSC_VER) && (defined _UNICODE))
+#ifdef _MSC_VER
 #define _tinydir_char_t TCHAR
-#define TINYDIR_STRING(s) TEXT(s)
+#define TINYDIR_STRING(s) _TEXT(s)
 #define _tinydir_strlen _tcslen
 #define _tinydir_strcpy _tcscpy
 #define _tinydir_strcat _tcscat
@@ -237,7 +237,7 @@ int tinydir_open(tinydir_dir *dir, const _tinydir_char_t *path)
 #ifdef _MSC_VER
 	_tinydir_strcpy(path_buf, dir->path);
 	_tinydir_strcat(path_buf, TINYDIR_STRING("\\*"));
-	dir->_h = FindFirstFileA(path_buf, &dir->_f);
+	dir->_h = FindFirstFile(path_buf, &dir->_f);
 	if (dir->_h == INVALID_HANDLE_VALUE)
 	{
 		errno = ENOENT;
@@ -388,7 +388,7 @@ int tinydir_next(tinydir_dir *dir)
 	}
 
 #ifdef _MSC_VER
-	if (FindNextFileA(dir->_h, &dir->_f) == 0)
+	if (FindNextFile(dir->_h, &dir->_f) == 0)
 #else
 #ifdef _TINYDIR_USE_READDIR
 	dir->_e = readdir(dir->_d);

--- a/tinydir.h
+++ b/tinydir.h
@@ -47,12 +47,15 @@ extern "C" {
 # include <sys/stat.h>
 # include <stddef.h>
 #endif
+#ifdef __MINGW32__
+# include <tchar.h>
+#endif
 
 
 /* types */
 
 /* Windows UNICODE wide character support */
-#ifdef _MSC_VER
+#if defined _MSC_VER || defined __MINGW32__
 #define _tinydir_char_t TCHAR
 #define TINYDIR_STRING(s) _TEXT(s)
 #define _tinydir_strlen _tcslen
@@ -113,6 +116,23 @@ extern "C" {
 # define _TINYDIR_USE_READDIR
 #endif
 
+/* MINGW32 has two versions of dirent, ASCII and UNICODE*/
+#ifndef _MSC_VER
+#if (defined __MINGW32__) && (defined _UNICODE)
+#define _TINYDIR_DIR _WDIR
+#define _tinydir_dirent _wdirent
+#define _tinydir_opendir _wopendir
+#define _tinydir_readdir _wreaddir
+#define _tinydir_closedir _wclosedir
+#else
+#define _TINYDIR_DIR DIR
+#define _tinydir_dirent dirent
+#define _tinydir_opendir opendir
+#define _tinydir_readdir readdir
+#define _tinydir_closedir closedir
+#endif
+#endif
+
 /* Allow user to use a custom allocator by defining _TINYDIR_MALLOC and _TINYDIR_FREE. */
 #if    defined(_TINYDIR_MALLOC) &&  defined(_TINYDIR_FREE)
 #elif !defined(_TINYDIR_MALLOC) && !defined(_TINYDIR_FREE)
@@ -134,7 +154,11 @@ typedef struct tinydir_file
 	int is_reg;
 
 #ifndef _MSC_VER
+#ifdef __MINGW32__
+	struct _stat _s;
+#else
 	struct stat _s;
+#endif
 #endif
 } tinydir_file;
 
@@ -149,10 +173,10 @@ typedef struct tinydir_dir
 	HANDLE _h;
 	WIN32_FIND_DATAA _f;
 #else
-	DIR *_d;
-	struct dirent *_e;
+	_TINYDIR_DIR *_d;
+	struct _tinydir_dirent *_e;
 #ifndef _TINYDIR_USE_READDIR
-	struct dirent *_ep;
+	struct _tinydir_dirent *_ep;
 #endif
 #endif
 } tinydir_dir;
@@ -183,7 +207,7 @@ int _tinydir_file_cmp(const void *a, const void *b);
 #ifndef _MSC_VER
 #ifndef _TINYDIR_USE_READDIR
 _TINYDIR_FUNC
-size_t _tinydir_dirent_buf_size(DIR *dirp);
+size_t _tinydir_dirent_buf_size(_TINYDIR_DIR *dirp);
 #endif
 #endif
 
@@ -242,7 +266,7 @@ int tinydir_open(tinydir_dir *dir, const _tinydir_char_t *path)
 	{
 		errno = ENOENT;
 #else
-	dir->_d = opendir(path);
+	dir->_d = _tinydir_opendir(path);
 	if (dir->_d == NULL)
 	{
 #endif
@@ -253,12 +277,12 @@ int tinydir_open(tinydir_dir *dir, const _tinydir_char_t *path)
 	dir->has_next = 1;
 #ifndef _MSC_VER
 #ifdef _TINYDIR_USE_READDIR
-	dir->_e = readdir(dir->_d);
+	dir->_e = _tinydir_readdir(dir->_d);
 #else
 	/* allocate dirent buffer for readdir_r */
 	size = _tinydir_dirent_buf_size(dir->_d); /* conversion to int */
 	if (size == -1) return -1;
-	dir->_ep = (struct dirent*)_TINYDIR_MALLOC(size);
+	dir->_ep = (struct _tinydir_dirent*)_TINYDIR_MALLOC(size);
 	if (dir->_ep == NULL) return -1;
 
 	error = readdir_r(dir->_d, dir->_ep, &dir->_e);
@@ -362,7 +386,7 @@ void tinydir_close(tinydir_dir *dir)
 #else
 	if (dir->_d)
 	{
-		closedir(dir->_d);
+		_tinydir_closedir(dir->_d);
 	}
 	dir->_d = NULL;
 	dir->_e = NULL;
@@ -391,7 +415,7 @@ int tinydir_next(tinydir_dir *dir)
 	if (FindNextFile(dir->_h, &dir->_f) == 0)
 #else
 #ifdef _TINYDIR_USE_READDIR
-	dir->_e = readdir(dir->_d);
+	dir->_e = _tinydir_readdir(dir->_d);
 #else
 	if (dir->_ep == NULL)
 	{
@@ -474,7 +498,12 @@ int tinydir_readfile(const tinydir_dir *dir, tinydir_file *file)
 	);
 	_tinydir_strcat(file->path, file->name);
 #ifndef _MSC_VER
-	if (stat(file->path, &file->_s) == -1)
+#ifdef __MINGW32__
+	if (_tstat(
+#else
+	if (stat(
+#endif
+		file->path, &file->_s) == -1)
 	{
 		return -1;
 	}
@@ -583,11 +612,7 @@ int tinydir_file_open(tinydir_file *file, const _tinydir_char_t *path)
 
 	/* Get the parent path */
 #ifdef _MSC_VER
-#ifdef _UNICODE
-        if (_wsplitpath_s(
-#else
-	if (_splitpath_s(
-#endif
+	if (_tsplitpath_s(
 			path,
 			drive_buf, sizeof drive_buf,
 			dir_name_buf, sizeof dir_name_buf,
@@ -605,9 +630,9 @@ int tinydir_file_open(tinydir_file *file, const _tinydir_char_t *path)
 	base_name = file_name_buf;
 #else
 	_tinydir_strcpy(dir_name_buf, path);
-	dir_name = dirname(dir_name_buf);
+	dir_name = TINYDIR_STRING(dirname(dir_name_buf));
 	_tinydir_strcpy(file_name_buf, path);
-	base_name = basename(file_name_buf);
+	base_name = TINYDIR_STRING(basename(file_name_buf));
 #endif
 
 	/* Open the parent directory */
@@ -683,7 +708,7 @@ from https://womble.decadent.org.uk/readdir_r-advisory.html
 * 255, since some systems (including at least HP-UX) incorrectly    *
 * define it to be a smaller value.                                  */
 _TINYDIR_FUNC
-size_t _tinydir_dirent_buf_size(DIR *dirp)
+size_t _tinydir_dirent_buf_size(_TINYDIR_DIR *dirp)
 {
 	long name_max;
 	size_t name_end;
@@ -703,9 +728,9 @@ size_t _tinydir_dirent_buf_size(DIR *dirp)
 #else
 #error "buffer size for readdir_r cannot be determined"
 #endif
-	name_end = (size_t)offsetof(struct dirent, d_name) + name_max + 1;
-	return (name_end > sizeof(struct dirent) ?
-		name_end : sizeof(struct dirent));
+	name_end = (size_t)offsetof(struct _tinydir_dirent, d_name) + name_max + 1;
+	return (name_end > sizeof(struct _tinydir_dirent) ?
+		name_end : sizeof(struct _tinydir_dirent));
 }
 #endif
 #endif

--- a/tinydir.h
+++ b/tinydir.h
@@ -38,6 +38,7 @@ extern "C" {
 #ifdef _MSC_VER
 # define WIN32_LEAN_AND_MEAN
 # include <windows.h>
+# include <tchar.h>
 # pragma warning(push)
 # pragma warning (disable : 4996)
 #else
@@ -49,6 +50,27 @@ extern "C" {
 
 
 /* types */
+
+/* Windows UNICODE wide character support */
+#if ((defined _MSC_VER) && (defined _UNICODE))
+#define _tinydir_char_t TCHAR
+#define TINYDIR_STRING(s) TEXT(s)
+#define _tinydir_strlen _tcslen
+#define _tinydir_strcpy _tcscpy
+#define _tinydir_strcat _tcscat
+#define _tinydir_strcmp _tcscmp
+#define _tinydir_strrchr _tcsrchr
+#define _tinydir_strncmp _tcsncmp
+#else
+#define _tinydir_char_t char
+#define TINYDIR_STRING(s) s
+#define _tinydir_strlen strlen
+#define _tinydir_strcpy strcpy
+#define _tinydir_strcat strcat
+#define _tinydir_strcmp strcmp
+#define _tinydir_strrchr strrchr
+#define _tinydir_strncmp strncmp
+#endif
 
 #define _TINYDIR_PATH_MAX 4096
 #ifdef _MSC_VER
@@ -105,9 +127,9 @@ extern "C" {
 
 typedef struct tinydir_file
 {
-	char path[_TINYDIR_PATH_MAX];
-	char name[_TINYDIR_FILENAME_MAX];
-	char *extension;
+	_tinydir_char_t path[_TINYDIR_PATH_MAX];
+	_tinydir_char_t name[_TINYDIR_FILENAME_MAX];
+	_tinydir_char_t *extension;
 	int is_dir;
 	int is_reg;
 
@@ -118,7 +140,7 @@ typedef struct tinydir_file
 
 typedef struct tinydir_dir
 {
-	char path[_TINYDIR_PATH_MAX];
+	_tinydir_char_t path[_TINYDIR_PATH_MAX];
 	int has_next;
 	size_t n_files;
 
@@ -139,9 +161,9 @@ typedef struct tinydir_dir
 /* declarations */
 
 _TINYDIR_FUNC
-int tinydir_open(tinydir_dir *dir, const char *path);
+int tinydir_open(tinydir_dir *dir, const _tinydir_char_t *path);
 _TINYDIR_FUNC
-int tinydir_open_sorted(tinydir_dir *dir, const char *path);
+int tinydir_open_sorted(tinydir_dir *dir, const _tinydir_char_t *path);
 _TINYDIR_FUNC
 void tinydir_close(tinydir_dir *dir);
 
@@ -169,7 +191,7 @@ size_t _tinydir_dirent_buf_size(DIR *dirp);
 /* definitions*/
 
 _TINYDIR_FUNC
-int tinydir_open(tinydir_dir *dir, const char *path)
+int tinydir_open(tinydir_dir *dir, const _tinydir_char_t *path)
 {
 #ifndef _MSC_VER
 #ifndef _TINYDIR_USE_READDIR
@@ -177,16 +199,16 @@ int tinydir_open(tinydir_dir *dir, const char *path)
 	int size;	/* using int size */
 #endif
 #else
-	char path_buf[_TINYDIR_PATH_MAX];
+	_tinydir_char_t path_buf[_TINYDIR_PATH_MAX];
 #endif
-	char *pathp;
+	_tinydir_char_t *pathp;
 
-	if (dir == NULL || path == NULL || strlen(path) == 0)
+	if (dir == NULL || path == NULL || _tinydir_strlen(path) == 0)
 	{
 		errno = EINVAL;
 		return -1;
 	}
-	if (strlen(path) + _TINYDIR_PATH_EXTRA >= _TINYDIR_PATH_MAX)
+	if (_tinydir_strlen(path) + _TINYDIR_PATH_EXTRA >= _TINYDIR_PATH_MAX)
 	{
 		errno = ENAMETOOLONG;
 		return -1;
@@ -204,17 +226,17 @@ int tinydir_open(tinydir_dir *dir, const char *path)
 #endif
 	tinydir_close(dir);
 
-	strcpy(dir->path, path);
+	_tinydir_strcpy(dir->path, path);
 	/* Remove trailing slashes */
-	pathp = &dir->path[strlen(dir->path) - 1];
-	while (pathp != dir->path && (*pathp == '\\' || *pathp == '/'))
+	pathp = &dir->path[_tinydir_strlen(dir->path) - 1];
+	while (pathp != dir->path && (*pathp == TINYDIR_STRING('\\') || *pathp == TINYDIR_STRING('/')))
 	{
-		*pathp = '\0';
+		*pathp = TINYDIR_STRING('\0');
 		pathp++;
 	}
 #ifdef _MSC_VER
-	strcpy(path_buf, dir->path);
-	strcat(path_buf, "\\*");
+	_tinydir_strcpy(path_buf, dir->path);
+	_tinydir_strcat(path_buf, TINYDIR_STRING("\\*"));
 	dir->_h = FindFirstFileA(path_buf, &dir->_f);
 	if (dir->_h == INVALID_HANDLE_VALUE)
 	{
@@ -256,7 +278,7 @@ bail:
 }
 
 _TINYDIR_FUNC
-int tinydir_open_sorted(tinydir_dir *dir, const char *path)
+int tinydir_open_sorted(tinydir_dir *dir, const _tinydir_char_t *path)
 {
 	/* Count the number of files first, to pre-allocate the files array */
 	size_t n_files = 0;
@@ -415,8 +437,8 @@ int tinydir_readfile(const tinydir_dir *dir, tinydir_file *file)
 		errno = ENOENT;
 		return -1;
 	}
-	if (strlen(dir->path) +
-		strlen(
+	if (_tinydir_strlen(dir->path) +
+		_tinydir_strlen(
 #ifdef _MSC_VER
 			dir->_f.cFileName
 #else
@@ -429,7 +451,7 @@ int tinydir_readfile(const tinydir_dir *dir, tinydir_file *file)
 		errno = ENAMETOOLONG;
 		return -1;
 	}
-	if (strlen(
+	if (_tinydir_strlen(
 #ifdef _MSC_VER
 			dir->_f.cFileName
 #else
@@ -441,16 +463,16 @@ int tinydir_readfile(const tinydir_dir *dir, tinydir_file *file)
 		return -1;
 	}
 
-	strcpy(file->path, dir->path);
-	strcat(file->path, "/");
-	strcpy(file->name,
+	_tinydir_strcpy(file->path, dir->path);
+	_tinydir_strcat(file->path, TINYDIR_STRING("/"));
+	_tinydir_strcpy(file->name,
 #ifdef _MSC_VER
 		dir->_f.cFileName
 #else
 		dir->_e->d_name
 #endif
 	);
-	strcat(file->path, file->name);
+	_tinydir_strcat(file->path, file->name);
 #ifndef _MSC_VER
 	if (stat(file->path, &file->_s) == -1)
 	{
@@ -510,7 +532,7 @@ int tinydir_readfile_n(const tinydir_dir *dir, tinydir_file *file, size_t i)
 _TINYDIR_FUNC
 int tinydir_open_subdir_n(tinydir_dir *dir, size_t i)
 {
-	char path[_TINYDIR_PATH_MAX];
+	_tinydir_char_t path[_TINYDIR_PATH_MAX];
 	if (dir == NULL)
 	{
 		errno = EINVAL;
@@ -522,7 +544,7 @@ int tinydir_open_subdir_n(tinydir_dir *dir, size_t i)
 		return -1;
 	}
 
-	strcpy(path, dir->_files[i].path);
+	_tinydir_strcpy(path, dir->_files[i].path);
 	tinydir_close(dir);
 	if (tinydir_open_sorted(dir, path) == -1)
 	{
@@ -534,26 +556,26 @@ int tinydir_open_subdir_n(tinydir_dir *dir, size_t i)
 
 /* Open a single file given its path */
 _TINYDIR_FUNC
-int tinydir_file_open(tinydir_file *file, const char *path)
+int tinydir_file_open(tinydir_file *file, const _tinydir_char_t *path)
 {
 	tinydir_dir dir;
 	int result = 0;
 	int found = 0;
-	char dir_name_buf[_TINYDIR_PATH_MAX];
-	char file_name_buf[_TINYDIR_FILENAME_MAX];
-	char *dir_name;
-	char *base_name;
+	_tinydir_char_t dir_name_buf[_TINYDIR_PATH_MAX];
+	_tinydir_char_t file_name_buf[_TINYDIR_FILENAME_MAX];
+	_tinydir_char_t *dir_name;
+	_tinydir_char_t *base_name;
 #ifdef _MSC_VER
-	char drive_buf[_TINYDIR_PATH_MAX];
-	char ext_buf[_TINYDIR_FILENAME_MAX];
+	_tinydir_char_t drive_buf[_TINYDIR_PATH_MAX];
+	_tinydir_char_t ext_buf[_TINYDIR_FILENAME_MAX];
 #endif
 
-	if (file == NULL || path == NULL || strlen(path) == 0)
+	if (file == NULL || path == NULL || _tinydir_strlen(path) == 0)
 	{
 		errno = EINVAL;
 		return -1;
 	}
-	if (strlen(path) + _TINYDIR_PATH_EXTRA >= _TINYDIR_PATH_MAX)
+	if (_tinydir_strlen(path) + _TINYDIR_PATH_EXTRA >= _TINYDIR_PATH_MAX)
 	{
 		errno = ENAMETOOLONG;
 		return -1;
@@ -561,7 +583,11 @@ int tinydir_file_open(tinydir_file *file, const char *path)
 
 	/* Get the parent path */
 #ifdef _MSC_VER
+#ifdef _UNICODE
+        if (_wsplitpath_s(
+#else
 	if (_splitpath_s(
+#endif
 			path,
 			drive_buf, sizeof drive_buf,
 			dir_name_buf, sizeof dir_name_buf,
@@ -572,15 +598,15 @@ int tinydir_file_open(tinydir_file *file, const char *path)
 		return -1;
 	}
 	/* Concatenate the drive letter and dir name to form full dir name */
-	strcat(drive_buf, dir_name_buf);
+	_tinydir_strcat(drive_buf, dir_name_buf);
 	dir_name = drive_buf;
 	/* Concatenate the file name and extension to form base name */
-	strcat(file_name_buf, ext_buf);
+	_tinydir_strcat(file_name_buf, ext_buf);
 	base_name = file_name_buf;
 #else
-	strcpy(dir_name_buf, path);
+	_tinydir_strcpy(dir_name_buf, path);
 	dir_name = dirname(dir_name_buf);
-	strcpy(file_name_buf, path);
+	_tinydir_strcpy(file_name_buf, path);
 	base_name = basename(file_name_buf);
 #endif
 
@@ -598,7 +624,7 @@ int tinydir_file_open(tinydir_file *file, const char *path)
 			result = -1;
 			goto bail;
 		}
-		if (strcmp(file->name, base_name) == 0)
+		if (_tinydir_strcmp(file->name, base_name) == 0)
 		{
 			/* File found */
 			found = 1;
@@ -620,10 +646,10 @@ bail:
 _TINYDIR_FUNC
 void _tinydir_get_ext(tinydir_file *file)
 {
-	char *period = strrchr(file->name, '.');
+	_tinydir_char_t *period = _tinydir_strrchr(file->name, TINYDIR_STRING('.'));
 	if (period == NULL)
 	{
-		file->extension = &(file->name[strlen(file->name)]);
+		file->extension = &(file->name[_tinydir_strlen(file->name)]);
 	}
 	else
 	{
@@ -640,7 +666,7 @@ int _tinydir_file_cmp(const void *a, const void *b)
 	{
 		return -(fa->is_dir - fb->is_dir);
 	}
-	return strncmp(fa->name, fb->name, _TINYDIR_FILENAME_MAX);
+	return _tinydir_strncmp(fa->name, fb->name, _TINYDIR_FILENAME_MAX);
 }
 
 #ifndef _MSC_VER

--- a/tinydir.h
+++ b/tinydir.h
@@ -594,7 +594,7 @@ int tinydir_file_open(tinydir_file *file, const _tinydir_char_t *path)
 	_tinydir_char_t file_name_buf[_TINYDIR_FILENAME_MAX];
 	_tinydir_char_t *dir_name;
 	_tinydir_char_t *base_name;
-#ifdef _MSC_VER
+#if (defined _MSC_VER || defined __MINGW32__)
 	_tinydir_char_t drive_buf[_TINYDIR_PATH_MAX];
 	_tinydir_char_t ext_buf[_TINYDIR_FILENAME_MAX];
 #endif
@@ -611,15 +611,25 @@ int tinydir_file_open(tinydir_file *file, const _tinydir_char_t *path)
 	}
 
 	/* Get the parent path */
+#if (defined _MSC_VER || defined __MINGW32__)
 #ifdef _MSC_VER
-	if (_tsplitpath_s(
+		_tsplitpath_s(
 			path,
 			drive_buf, sizeof drive_buf,
 			dir_name_buf, sizeof dir_name_buf,
 			file_name_buf, sizeof file_name_buf,
-			ext_buf, sizeof ext_buf))
+			ext_buf, sizeof ext_buf);
+#else
+		_tsplitpath(
+			path,
+			drive_buf,
+			dir_name_buf,
+			file_name_buf,
+			ext_buf);
+#endif
+	if (errno)
 	{
-		errno = EINVAL;
+//		errno = EINVAL;
 		return -1;
 	}
 	/* Concatenate the drive letter and dir name to form full dir name */
@@ -630,9 +640,9 @@ int tinydir_file_open(tinydir_file *file, const _tinydir_char_t *path)
 	base_name = file_name_buf;
 #else
 	_tinydir_strcpy(dir_name_buf, path);
-	dir_name = TINYDIR_STRING(dirname(dir_name_buf));
+	dir_name = dirname(dir_name_buf);
 	_tinydir_strcpy(file_name_buf, path);
-	base_name = TINYDIR_STRING(basename(file_name_buf));
+	base_name =basename(file_name_buf);
 #endif
 
 	/* Open the parent directory */

--- a/tinydir.h
+++ b/tinydir.h
@@ -171,7 +171,7 @@ typedef struct tinydir_dir
 	tinydir_file *_files;
 #ifdef _MSC_VER
 	HANDLE _h;
-	WIN32_FIND_DATAA _f;
+	WIN32_FIND_DATA _f;
 #else
 	_TINYDIR_DIR *_d;
 	struct _tinydir_dirent *_e;

--- a/tinydir.h
+++ b/tinydir.h
@@ -629,7 +629,7 @@ int tinydir_file_open(tinydir_file *file, const _tinydir_char_t *path)
 #endif
 	if (errno)
 	{
-//		errno = EINVAL;
+		errno = EINVAL;
 		return -1;
 	}
 	/* Concatenate the drive letter and dir name to form full dir name */

--- a/tinydir.h
+++ b/tinydir.h
@@ -612,7 +612,7 @@ int tinydir_file_open(tinydir_file *file, const _tinydir_char_t *path)
 
 	/* Get the parent path */
 #if (defined _MSC_VER || defined __MINGW32__)
-#ifdef _MSC_VER
+#if ((defined _MSC_VER) && (_MSC_VER >= 1400))
 		_tsplitpath_s(
 			path,
 			drive_buf, sizeof drive_buf,

